### PR TITLE
Better mitigation of share inflation

### PR DESCRIPTION
**Pull request by ChaoticWalrus**
See commits. Primary change is to StrategyBase; everything else is to tests.
In reviewing you should specifically (in addition to normal review):

- read through and understand the OZ issue thread I've linked 
- verify that all of the comments I've added are sufficient explanation
- consider and understand the impact of the chosen values of `SHARES_OFFSET` and `BALANCE_OFFSET`
- comment if you think the chosen values are optimal. why / why not

### CODE REVIEW

File: src/test/DepositWithdraw.t.sol [hunk 1]
```diff
@@ -400,25 +396,73 @@ contract DepositWithdrawTests is EigenLayerTestHelper {
...
+        require(attackerCost >= attackerValueWeth, "attacker gained value?");
+        // uint256 attackerLossesWeth = attackerValueWeth > attackerCost ? 0 : (attackerCost - attackerValueWeth);
+        uint256 attackerLossesWeth = attackerCost - attackerValueWeth;
+        uint256 userLossesWeth = secondDepositAmount - userValueWeth;
+
+        emit log_named_uint("attackerLossesWeth", attackerLossesWeth);
+        emit log_named_uint("userLossesWeth", userLossesWeth);
+
+        // use '+1' here to account for rounding. given the attack will cost ETH in the form of gas, this is fine.
+        require(attackerLossesWeth + 1 >= userLossesWeth, "griefing attack deals more damage than cost");
```
Comment (gpsanant) [inline src/test/DepositWithdraw.t.sol:451]: :goat: test

File: src/test/DepositWithdraw.t.sol [hunk 2]
```diff
@@ -553,8 +643,49 @@ contract DepositWithdrawTests is EigenLayerTestHelper {
...
+                "_testDepositToStrategy: stakerStrategyList array updated incorrectly"
+            );
+        }
+            
+        // check that the shares out match the expected amount out
+        // the actual transfer in will be lower by 1 wei than expected due to stETH's internal rounding
+        // to account for this we check approximate rather than strict equivalence here
+        {
+            uint256 actualSharesOut = strategyManager.stakerStrategyShares(address(this), stethStrategy) - operatorSharesBefore;
+            require((actualSharesOut * 1000) / expectedSharesOut > 998, "too few shares");
```
Comment (gpsanant) [inline src/test/DepositWithdraw.t.sol:682]: shouldn't this slippage be way less?

Comment (ChaoticWalrus) [reply src/test/DepositWithdraw.t.sol]: re-check with the changes in https://github.com/Layr-Labs/eigenlayer-contracts/pull/36/commits/e8cc3f9979071b22c6b121500abfc6e17faa4443

**Review by gpsanant — APPROVED** [review]
:+1:

**Pull request merged**